integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonLoadDataCommand.scala

```diff
@@ -345,23 +380,172 @@ case class CarbonLoadDataCommand(
+        dateFormatString = carbonLoadModel.getDefaultDateFormat
+      }
+      val dateFormat = new SimpleDateFormat(dateFormatString)
+      // input data from csv files. Convert to logical plan
+      CommonUtil.configureCSVInputFormat(hadoopConf, carbonLoadModel)
+      hadoopConf.set(FileInputFormat.INPUT_DIR, carbonLoadModel.getFactFilePath)
+      val jobConf = new JobConf(hadoopConf)
+      SparkHadoopUtil.get.addCredentials(jobConf)
+      val attributes =
+        StructType(carbonLoadModel.getCsvHeaderColumns.map(
+          StructField(_, StringType))).toAttributes
+      val rowDataTypes = attributes.map{f =>
+        relation.output.find(_.name.equalsIgnoreCase(f.name)) match {
+          case Some(attr) => attr.dataType
+          case _ => StringType
+        }
+      }
+      val len = rowDataTypes.length
+      val rdd =
+        new NewHadoopRDD[NullWritable, StringArrayWritable](
+          sparkSession.sparkContext,
+          classOf[CSVInputFormat],
+          classOf[NullWritable],
+          classOf[StringArrayWritable],
+          jobConf).map{f =>
+            val data = new Array[Any](len)
+            var i = 0
+            while (i < len) {
+              // TODO find a way to avoid double conversion of date and time.
+              data(i) = CarbonScalaUtil.getString(
+                f._2.get()(i),
+                rowDataTypes(i),
+                timeStampFormat,
+                dateFormat)
+              i = i + 1
+            }
+            InternalRow.fromSeq(data)
+        }
+
+      // Only select the required columns
+      Project(relation.output.map(f => attributes.find(_.name.equalsIgnoreCase(f.name)).get),
+        LogicalRDD(attributes, rdd)(sparkSession))
+    }
+    Dataset.ofRows(sparkSession, InsertIntoTable(
+      convertToLogicalRelation(relation, isOverwriteTable, carbonLoadModel, sparkSession),
+      partition,
+      query,
+      OverwriteOptions(isOverwriteTable), false))
+  }
+
+  private def convertToLogicalRelation(
+      relation: LogicalRelation,
+      overWrite: Boolean,
+      loadModel: CarbonLoadModel,
+      sparkSession: SparkSession): LogicalRelation = {
+    val catalogTable = relation.catalogTable.get
+    val table = loadModel.getCarbonDataLoadSchema.getCarbonTable
+    val metastoreSchema = StructType(StructType.fromAttributes(
+      relation.output).fields.map(_.copy(dataType = StringType)))
+    val lazyPruningEnabled = sparkSession.sqlContext.conf.manageFilesourcePartitions
+    val catalog = new CatalogFileIndex(
+      sparkSession, catalogTable, relation.relation.sizeInBytes)
+    if (lazyPruningEnabled) {
+      catalog
+    } else {
+      catalog.filterPartitions(Nil) // materialize all the partitions in memory
+    }
+    val partitionSchema =
+      StructType(table.getPartitionInfo(table.getTableName).getColumnSchemaList.asScala.map(f =>
+      metastoreSchema.fields.find(_.name.equalsIgnoreCase(f.getColumnName))).map(_.get))
```

Comment:
can you make it more readable, it is not easy to check the map(_.get)

Reply:
ok